crates/sui-core/src/authority/authority_store_pruner.rs

```diff
@@ -0,0 +1,302 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{cmp::Ordering, sync::Arc, time::Duration};
+use sui_config::node::AuthorityStorePruningConfig;
+use sui_types::base_types::{ObjectID, SequenceNumber, VersionNumber};
+use tokio::{
+    sync::oneshot::{self, Sender},
+    time::{self, Instant},
+};
+use tracing::log::{error, info};
+use typed_store::Map;
+
+use crate::authority::authority_store::ObjectKey;
+
+use super::authority_store_tables::AuthorityPerpetualTables;
+
+const MAX_OPS_IN_ONE_WRITE_BATCH: u64 = 10000;
+
+pub struct AuthorityStorePruner {
+    _objects_pruner_cancel_handle: oneshot::Sender<()>,
+}
+
+impl AuthorityStorePruner {
+    fn prune_objects(
+        num_versions_to_retain: u64,
+        perpetual_db: Arc<AuthorityPerpetualTables>,
+    ) -> u64 {
+        let iter = perpetual_db.objects.iter().skip_to_last().reverse();
+        let mut total_keys_scanned = 0;
+        let mut total_objects_scanned = 0;
+        let mut num_versions_for_key = 0;
+        let mut object_id: Option<ObjectID> = None;
+        let mut seq_num: Option<VersionNumber> = None;
+        let mut num_pending_wb_ops = 0;
+        let mut total_pruned: u64 = 0;
+        let mut wb = perpetual_db.objects.batch();
+        for (key, _value) in iter {
```

Comment:
I had assumed that we would do pruning incrementally, after execution. So, when we finish a transaction, we have a set of objects that are now candidates for pruning. Alternatively we could post-process effects in larger batches, perhaps grouped by checkpoint.

This will work for now, but a full table scan is not going to scale very well, especially if we assume object usage will follow an 80/20 power law where the large majority of objects change very rarely.

Reply:
Thanks @mystenmark, those are good points to bring up. The reason to not continually delete objects after transaction execution is that it will leave a lot of delete markers (tombstones) in the database which is known to make iterate really slow (we use iter() to get latest version of an object for example) and hurt performance. We can be smart by invoking pruning on hot ranges (key prefixes which are seeing a lot of new object versions being added) at higher frequency (where the data for which key ranges are hot or which object ids need pruning can be cheaply collected+retained during transaction execution) i.e. every few hours and pruning on entire table at lower frequency i.e. once a day perhaps. Right now in tests the table size stayed well below few hundred MBs even at > 200tps and hence I did not decide to build anything else but happy to course correct and optimize this if it turns out to be a problem.

Reply:
Do we have any tests that actually grow the live set though? For instance, I believe the stress test never creates net new objects (excluding the initial gas object creation phase of course).

I hear what you're saying about avoiding on-the-spot deletions. This is why I mentioned batching up a lot of work at once using checkpoints. We can even do an entire epoch at a time. My point is not to do the deletions immediately, but to avoid doing a full table scan - TransactionEffects contain all the info we need to know exactly which objects (even which versions) need to be pruned.

We don't need to tackle this now, but lets keep this in mind as something we might need to tackle urgently at mainnet, depending on usage patterns.

Reply:
Our stress tests don't create net new objects but they continuously create new versions of existing objects i.e. gas objects, other input objects, etc 
You have a good point about leveraging transaction effects for this where reading effects is proportional to the volume of transaction throughput vs the current approach of scanning objects which is proportional to size of existing objects table. Although with reading checkpoint/effects approach, the tricky bit is knowing the right object version to prune upto. For example: if I read last 10 checkpoints, and find 2 latest versions for an object X and we want to retain 5 latest versions - then somehow we have to carry over this state to the next time we read another set of checkpoints. I may not have communicated what I am thinking about the corner cases very clearly here but we can definitely take it offline.
We may actually have to worry about this even less once object table [optimizations](https://docs.google.com/document/d/18DTsNZiSBudI-KIkb7Xky3wNgnCeAeq73KrptXq3vxo/edit#) are in which makes entries in the object table references (and scanning objects table becomes much cheaper because it has key -> pointer entries) and pruning of object content table is based off of reference counts.

Reply:
Btw, I agree that if for some unforeseen reason the above data structure optimizations cannot make it to mainnet then we should definitely explore the option you mentioned i.e. use checkpoint/effects based object ids to trigger delete on.

Reply:
> Our stress tests don't create net new objects but they continuously create new versions of existing objects i.e. gas objects, other input objects, etc

Yes,  exactly - this is the crucial point. Under the stress test, every single object ID may have many prunable old versions, and the number of unique Object IDs does not grow over time. So a table scan is not inefficient at all.

However, a real network will not look like that. First, the set of Object IDs will grow over time. Second, if we prune at any reasonably short interval, its quite likely that a large % of the objects won't have been written since the previous prune. So we will do the work of scanning those objects, but gain no benefit from doing so.

> Although with reading checkpoint/effects approach, the tricky bit is knowing the right object version to prune upto. For example: if I read last 10 checkpoints, and find 2 latest versions for an object X and we want to retain 5 latest versions - then somehow we have to carry over this state to the next time we read another set of checkpoints. 

Yes, that is a good point - probably the simplest thing is to just collect all the ObjectIDs that were touched by a set of effects, and then prune all prunable versions of those IDs, without paying attention to the IDs that were in the effects.

Reply:
> Yes, exactly - this is the crucial point. Under the stress test, every single object ID may have many prunable old versions, and the number of unique Object IDs does not grow over time. So a table scan is not inefficient at all.

Yeah, I completely agree our stress workload is not representative from pruning perspective. And your point makes sense.

> Yes, that is a good point - probably the simplest thing is to just collect all the ObjectIDs that were touched by a set of effects, and then prune all prunable versions of those IDs, without paying attention to the IDs that were in the effects.

I think this approach works well if the total number of object IDs to prune are such that cost (cpu) with iter.seek() does not dominate. For example - if we find 1M object ids in last 10 checkpoints, we will do 1M iter.seek() where each one of them does a separate binary search across all sorted runs (each run has many files) in the lsm tree to position the iterator. But since we are pruning in the background and on a single task, I don't think we can overload system resources. I think I am convinced about having such an incremental pruning system before mainnet. We however need to ensure incremental pruning is able to keep up with the current checkpoint generation rate and if it falls behind (and db size grows) then invoke full pruning.
I think we can let this PR for full pruning go for now but we can make it run at a lower frequency, perhaps even once per epoch. This will at least tell us the efficacy of pruning on real world workloads once this lands in devnet. And work on incremental pruning in the meantime. Let me know what you think.

Reply:
An mixed approach would be to sort the 1M objects and seek/skip to these keys only so ignoring the cold objects when pruning.

Reply:
Yeah we can take such an approach where we only look at the hot objects (creating new versions fast) if pruning realizes that it is lagging behind. And depend on full pruning to prune cold objects eventually.